src/Microsoft.OData.Core/Json/ODataJsonWriterUtilsAsync.cs

```diff
@@ -0,0 +1,274 @@
+        /// </summary>
+        /// <param name="jsonWriter">JSON writer.</param>
+        /// <param name="code">The code of the error.</param>
+        /// <param name="message">The message of the error.</param>
+        /// <param name="target">The target of the error.</param>
+        /// <param name="details">The details of the error.</param>
+        /// <param name="innerError">Inner error details that will be included in debug mode (if present).</param>
+        /// <param name="instanceAnnotations">Instance annotations for this error.</param>
+        /// <param name="writeInstanceAnnotationsDelegate">Delegate to write the instance annotations.</param>
+        /// <param name="maxInnerErrorDepth">The maximum number of nested inner errors to allow.</param>
+        /// <param name="writingJsonLight">true if we're writing JSON lite, false if we're writing verbose JSON.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        private static async Task WriteErrorAsync(IJsonWriterAsync jsonWriter, string code, string message, string target,
+            IEnumerable<ODataErrorDetail> details,
+            ODataInnerError innerError,
+            IEnumerable<ODataInstanceAnnotation> instanceAnnotations,
+            Func<IEnumerable<ODataInstanceAnnotation>, Task> writeInstanceAnnotationsDelegate, int maxInnerErrorDepth,
+            bool writingJsonLight)
+        {
+            Debug.Assert(jsonWriter != null, "jsonWriter != null");
+            Debug.Assert(code != null, "code != null");
+            Debug.Assert(message != null, "message != null");
+            Debug.Assert(instanceAnnotations != null, "instanceAnnotations != null");
+
+            // "error": {
+            await jsonWriter.StartObjectScopeAsync().ConfigureAwait(false);
+            if (writingJsonLight)
+            {
+                await jsonWriter.WriteNameAsync(JsonLightConstants.ODataErrorPropertyName).ConfigureAwait(false);
+            }
+            else
+            {
+                await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorName).ConfigureAwait(false);
+            }
+
+            await jsonWriter.StartObjectScopeAsync().ConfigureAwait(false);
+
+            // "code": "<code>"
+            await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorCodeName).ConfigureAwait(false);
+            await jsonWriter.WriteValueAsync(code).ConfigureAwait(false);
+
+            // "message": "<message string>"
+            await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorMessageName).ConfigureAwait(false);
+            await jsonWriter.WriteValueAsync(message).ConfigureAwait(false);
+
+            // For example, "target": "query",
+            if (target != null)
+            {
+                await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorTargetName).ConfigureAwait(false);
+                await jsonWriter.WriteValueAsync(target).ConfigureAwait(false);

```

Comment:
the old OData error payload is not same as the format defined in the spec. If possible, should we fix it in the new API?